embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+/// will return a `SendError`. Similarly, if `Sender` is disconnected while
+/// trying to `recv`, the `recv` method will return a `RecvError`.
+///
+/// Note that when splitting the channel, the sender and receiver cannot outlive
+/// their channel. The following will therefore fail compilation:
+////
+/// ```compile_fail
+/// use embassy::util::mpsc;
+/// use embassy::util::mpsc::{Channel, WithThreadModeOnly};
+///
+/// let (sender, receiver) = {
+///    let mut channel = Channel::<WithThreadModeOnly, u32, 3>::with_thread_mode_only();
+///     mpsc::split(&channel)
+/// };
+/// ```
+pub fn split<M, T, const N: usize>(
+    channel: &Channel<M, T, N>,
+) -> (Sender<M, T, N>, Receiver<M, T, N>)
+where
+    M: Mutex<Data = ()>,
+{
+    let sender = Sender { channel: &channel };
+    let receiver = Receiver { channel: &channel };
+    {
+        let c = channel.get();
+        c.register_receiver();
+        c.register_sender();
+    }
+    (sender, receiver)
+}
+
+impl<'ch, M, T, const N: usize> Receiver<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    /// Receives the next value for this receiver.
+    ///
+    /// This method returns `None` if the channel has been closed and there are
+    /// no remaining messages in the channel's buffer. This indicates that no
+    /// further values can ever be received from this `Receiver`. The channel is
+    /// closed when all senders have been dropped, or when [`close`] is called.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.
+    ///
+    /// Note that if [`close`] is called, but there are still outstanding
+    /// messages from before it was closed, the channel is not considered
+    /// closed by `recv` until they are all consumed.
+    ///
+    /// [`close`]: Self::close
+    pub async fn recv(&mut self) -> Option<T> {
+        self.await
+    }
+
+    /// Attempts to immediately receive a message on this `Receiver`
+    ///
+    /// This method will either receive a message from the channel immediately or return an error
+    /// if the channel is empty.
+    pub fn try_recv(&self) -> Result<T, TryRecvError> {
+        self.channel.get().try_recv()
+    }
+
+    /// Closes the receiving half of a channel without dropping it.
+    ///
+    /// This prevents any further messages from being sent on the channel while
+    /// still enabling the receiver to drain messages that are buffered.
+    ///
+    /// To guarantee that no messages are dropped, after calling `close()`,
+    /// `recv()` must be called until `None` is returned. If there are
+    /// outstanding messages, the `recv` method will not return `None`
+    /// until those are released.
+    ///
+    pub fn close(&mut self) {
+        self.channel.get().close()
+    }
+}
+
+impl<'ch, M, T, const N: usize> Future for Receiver<'ch, M, T, N>
```

Comment:
It's somewhat strange that `Receiver` impls `Future`. Futures are only supposed to run once. Maybe making a `ReceiveFuture` would make sense.

Reply:
> It's somewhat strange that `Receiver` impls `Future`. 

There are other examples of this in Embassy, e.g. `Timer`.

> Futures are only supposed to run once. Maybe making a `ReceiveFuture` would make sense.

Whoops - mistake. Commit ec18ae1f8836815773a5ba744c65404bc64f0af4 fixes this.